### PR TITLE
ArrayFire bindings for Linear Algebra

### PR DESCRIPTION
**Pull request by prasunanand**
Implemented following classes:
1. Af_Array
2. Algorithm
3. Arith
4. BLAS
5. LAPACK
6. CUDA 
7. OpenCL
8. Device
9. Random
10. Sparse
11. Statistics

Added exception handling.
Only doubles implemented currently

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -21,3 +24,16 @@ end
 def run(*cmd)
   sh(cmd.join(' '))
 end
+
+RDoc::Task.new do |rdoc|
+  rdoc.main = "README.md"
+  rdoc.rdoc_files.include(%w{README.md LICENSE CONTRIBUTING.md lib ext})
+end
```
Comment (v0dro) [inline Rakefile:31]: Will this task run if you do a `gem build` on the gemspec? Is it not possible to add this to the gemspec itself?

File: ext/mri/cmodules/arith.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  afstruct* right;                                                         \
+  afstruct* result = ALLOC(afstruct);                                      \
+  Data_Get_Struct(left_val, afstruct, left);                               \
+  Data_Get_Struct(right_val, afstruct, right);                             \
+  af_##oper(&result->carray,  left->carray, right->carray, true);          \
+  af_print_array(result->carray);                                          \
+  return Data_Wrap_Struct(CLASS_OF(left_val), NULL, arf_free, result);     \
+}
+
+#define DEF_UNARY_RUBY_ACCESSOR(oper, name)                                \
```
Comment (v0dro) [inline ext/mri/cmodules/arith.c]: Nice use of macros!

File: Rakefile (hunk 1 above)
Comment (prasunanand) [reply Rakefile:31]: @v0dro I didn't get the question. Why would we need `rdoc` to run when using `gem build`.?

AFAIK, RubyDoc.info automatically generates the docs when I push the gem.

File: ext/mri/cmodules/device.c [hunk 3]
```diff
@@ -14,9 +14,19 @@ static VALUE arf_info_string(VALUE self, VALUE bool_val){
   return rb_str_new_cstr(str);
 }
 
-static VALUE arf_device_info(VALUE self){
-  char d_name, d_platform, d_toolkit, d_compute;
-  af_device_info(&d_name, &d_platform, &d_toolkit, &d_compute);
+static VALUE arf_device_info(VALUE self, VALUE name_val, VALUE platform_val, VALUE toolkit_val, VALUE compute_val){
+  char* d_name     = (char*)malloc(sizeof(char) * 64);
+  char* d_platform = (char*)malloc(sizeof(char) * 10);
+  char* d_toolkit  = (char*)malloc(sizeof(char) * 64);
```
Comment (v0dro) [inline ext/mri/cmodules/device.c]: Shouldn't you be using `xmalloc` and `xfree` everywhere? They're the functions from ruby.h and allocate memory that the ruby VM is aware. malloc will allocate from system heap not ruby VM heap.

File: Rakefile (hunk 1 above)
Comment (v0dro) [reply Rakefile:31]: So the above rake commands adds rdoc files which are then read by rubygems.org to generate YARD docs. If its not done during gem build and if you forget to run above rake task, the rdoc files will not get added to the gemspec and rubygems wont be able to detect them.

**Comment by v0dro** [comment]
When do you plan to merge this?

File: ext/mri/interfaces/nmatrix.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  Data_Get_Struct(self, afstruct, input);
+  dim_t count;
+  uint ndims;
+  af_get_numdims(&ndims, input->carray);
+
+  dim_t* dims = (dim_t*)malloc(ndims * sizeof(dim_t));
+
+  af_get_dims(&dims[0], &dims[1], &dims[2], &dims[3], input->carray);
+
+  size_t* shape = (size_t*)malloc(ndims * sizeof(size_t));;
```
Comment (v0dro) [inline ext/mri/interfaces/nmatrix.c]: Double colon at the end.

File: ext/mri/interfaces/nmatrix.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
+static VALUE arf_af_array_to_nmatrix(VALUE self) {
+  afstruct* input;
+  Data_Get_Struct(self, afstruct, input);
+  dim_t count;
+  uint ndims;
+  af_get_numdims(&ndims, input->carray);
+
+  dim_t* dims = (dim_t*)malloc(ndims * sizeof(dim_t));
```
Comment (v0dro) [inline ext/mri/interfaces/nmatrix.c]: its good practice in C to declare all the variables at the top of the function.

File: ext/mri/interfaces/nmatrix.c (hunk 4 above)
Comment (prasunanand) [reply ext/mri/interfaces/nmatrix.c]: Its outdated commit.

File: Rakefile (hunk 1 above)
Comment (prasunanand) [reply Rakefile:31]: @v0dro, I will follow Victor's advice about using `rubygems-tasks` library.

**Comment by prasunanand** [comment]
@9prady9 You can merge this PR now :) .

**Comment by v0dro** [comment]
Is it possible to publish this using rubygems? Just `gem install arrayfire-rb` would be amazing.

**Comment by v0dro** [comment]
Also you should remove the statement saying that bindings aren't yet ready for production from README. And, add SciRuby to the acknowledgements at the bottom along with the website.

**Comment by 9prady9** [comment]
@v0dro 
I am not aware of `rubygems`, but in principle it should be possible. We have similar packaging in rust environment(rust package manager - cargo.io) known as `crate`. From the usage style, it seems like rubygems is similar to crates in Rust or javascript package manager - npm.

I don't think it is a good idea to remove the cautionary statements until all the feasible functionality from ArrayFire is ported over to the ruby bindings. For one, I don't think @prasunanand has worked on image processing functions yet. Also, I believe indexing hasn't been worked on either.

**Comment by v0dro** [comment]
Oh. Ok. @prasunanand can you create issues and set milestones for this purpose in the GitHub issue tracker? It will becomes easy for new contributors to quickly what's going on and how to contribute. Make things as granular as possible.

@9prady9 rubygems is the ruby packaging system and the central repository for most open source gems (yes, it is similar to npm and crates). Hosting the gem on rubygems will allow users to install with a single command instead of cloning and compiling.

**Comment by 9prady9** [comment]
@prasunanand @v0dro Please use [this issue](https://github.com/arrayfire/arrayfire-rb/issues/4) to track progress of implementing left over features.

**Comment by prasunanand** [comment]
Thanks @9prady9 for merging the PR. 

@v0dro I am creating milestones and I will work on the missing features.

**Comment by v0dro** [comment]
Alright. I think you add them to @9prady9 's list.

**Pull request merged**